Classes/Private/Manager/Objects/OCTActiveFile.m

```diff
@@ -0,0 +1,504 @@
+            [self sendProgressUpdateNow];
+        });
+    }
+}
+
+- (void)wipeCounters
+{
+    @synchronized(self) {
+        for (int i = 0; i < kSecondsToAverage; ++i) {
+            self.transferRateCounters[i] = -1;
+        }
+    }
+    unsigned long t = OCTGetMonotonicTime();
+    self.lastCountedTime = t;
+    self.lastProgressUpdateTime = t;
+}
+
+#pragma mark - Private API
+
+- (NSData *)archiveConduit
+{
+    if ([self.conduit conformsToProtocol:@protocol(NSCoding)]) {
+        return [NSKeyedArchiver archivedDataWithRootObject:self.conduit];
+    }
+    return nil;
+}
+
+- (void)control:(OCTToxFileControl)ctl
+{
+    OCTMessageFile *mf = (OCTMessageFile *)[[self.fileManager.dataSource managerGetRealmManager] objectWithUniqueIdentifier:self.fileIdentifier class:[OCTMessageFile class]];
+
+    switch (ctl) {
+        case OCTToxFileControlCancel: {
+            DDLogDebug(@"_control: obeying cancel message from remote.");
+            [self stopFileNow];
+            [self markFileAsCancelled:self];
+            break;
+        }
+        case OCTToxFileControlPause:
+            DDLogDebug(@"_control: obeying pause message from remote.");
+            [self markFileAsPaused:self withFlags:mf.pauseFlags | OCTPauseFlagsFriend];
+            break;
+        case OCTToxFileControlResume: {
+            DDLogDebug(@"_control: obeying resume message from remote.");
+            if (mf.pauseFlags == OCTPauseFlagsFriend) {
+                dispatch_sync(self.fileManager.queue, ^{
+                    [self openConduitIfNeeded];
+                });
+                [self resumeFile:self];
+            }
+            else {
+                [self markFileAsPaused:self withFlags:OCTPauseFlagsSelf];
+            }
+            break;
+        }
+    }
+}
+
+- (void)interrupt
+{
+    [self stopFileNow];
+    [self markFileAsInterrupted:self];
+}
+
+#pragma mark - Internal API
+
+- (void)stopFileNow
+{
+    dispatch_async(self.fileManager.queue, ^{
+        [self closeConduitIfNeeded];
+
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self.fileManager removeFile:self];
+        });
+    });
+}
+
+- (id<OCTFileConduit>)conduit
+{
+    @throw [NSException exceptionWithName:@"OCTFileException" reason:@"Only subclasses of OCTActiveFile can be used." userInfo:nil];
+    return nil;
+}
+
+- (BOOL)openConduitIfNeeded
+{
+    if (! self.isConduitOpen) {
+        BOOL ret = [self.conduit transferWillBecomeActive:self];
+
+        if ([self.conduit respondsToSelector:@selector(moveToPosition:)]) {
+            [self.conduit moveToPosition:self.bytesMoved];
+        }
+
+        self.isConduitOpen = ret;
+        return ret;
+    }
+    else {
+        return YES;
+    }
+}
+
+- (void)closeConduitIfNeeded
+{
+    if (! self.isConduitOpen) {
+        return;
+    }
+    else {
+        [self.conduit transferWillBecomeInactive:self];
+        self.isConduitOpen = NO;
+    }
+}
+
+- (void)sendProgressUpdateNow
```

Comment:
And `dispatch_async` can go here too as notification always has to be called on main thread.